cmd/kubeadm/app/util/tokens_test.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+)
+
+func TestUsingEmptyTokenFails(t *testing.T) {
+	// Simulates what happens when you omit --token on the CLI
+	s := newSecretsWithToken("")
+
+	ok, err := UseGivenTokenIfValid(s)
+	assert.False(t, ok)
+	assert.NoError(t, err)
```

Comment:
I want to preface this with this is just debating opinions here. I think we both agree the error message needs improving, and that's all I really want. What follows is probably a bit of a rant.

> I don't think I agree with the overall sentiment about assertion libraries. I can absolutely add better messaging for these assertions, and you're right to call that out, but if you have a few things to assert in a test, I think Go's if statements (with required curly braces) make it far less semantically compact to indicate their function to the reader of the code.

Being compact does not necessarily make it easier for the reader, especially when the `if/else` checks in the test would more closely resemble the code under test with all the 

```go
if err != nil {
	// handle error
}
```

error handling. And honestly it's only saving 2 lines of code. Moreover, the goal should be that the error message is descriptive enough that I shouldn't have to read the test code too closely to determine what went wrong.

> Especially functions like "assert.Equals" can remove a lot of the boilerplate code associated with not only checking equality, but printing a nicely-formatted message including the expected vs. got for the variables instead of needing to construct the format yourself by hand.

This is what `assert.False` prints by default:

```console
	Error Trace:	tokens_test.go:31
	Error:		Should be false
```

You can add a message to make it something like this:

```console
	Error Trace:	tokens_test.go:31
	Error:		Should be false
	Messages:	UseGivenTokenIfValid indicated an empty token is valid
```

That's very verbose and mostly useless. Compare that to

```go
if got := UseGivenTokenIfValid(s); got != want {
	test.Errorf("UseGivenTokenIfValid indicated an empty token is valid: returned true, expected false")
}
```

```console
	tokens_test.go:31: UseGivenTokenIfValid indicated an empty token is valid: returned true, expected false
```

Concise one-line output. As a side note: this also a standard format for printing errors that most editors understand by default and can easily navigate to the line the test failed on. (That format being a line starting with file_name:line_no)

And then there's the complexity these libraries tend towards. I think it's best summed up here: https://golang.org/doc/faq#testing_framework

> A related point is that testing frameworks tend to develop into mini-languages of their own, with conditionals and controls and printing mechanisms, but Go already has all those capabilities; why recreate them? We'd rather write tests in Go; it's one fewer language to learn and the approach keeps the tests straightforward and easy to understand.

> Since this personal preference can easily swing both ways, can we put a ruling into the Coding Conventions doc? If the project preference is already documented somewhere, then sorry for missing it.

I thought we already fell back to the go standards, but I see we only call out Effective Go. But overall, this is something I'm more than willing to debate, but I do not want to mandate it.


TL;DR make the messages better. Be aware of the different ways of doing these things and the various trade-offs.